support slot batch for CLUSTERX SETSLOT

infdahai opened the pull request:
closes #1412. refer to #529

1. add stdint.h for uint_t type
 ref(https://godbolt.org/z/5q6zj5W9s)
2. add cluster_setslot_batch feature.
3. tests for parsedSlotRange(cpp unit and go test)
4. move parse func to commandX

torwig commented on src/cluster/cluster.cc:
```diff
@@ -93,32 +93,27 @@ Status Cluster::SetNodeId(const std::string &node_id) {
-// slot distribution and there are 16384 slot in our design.
-//
+// Set the slots to the node if new version is current version +1. It is useful
+// when we scale cluster avoid too many big messages. By th way,there are 16384 slots
```
Typo in the phrase "By the way". You can omit it entirely because it doesn't add value to the sentence or explanation.

torwig commented on src/cluster/cluster.cc:
```diff
@@ -128,30 +123,43 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
   }
 
+  std::vector<std::pair<int, int>> slots;
+  Status s = parseSlotRange(slots_str, slots);
```
Is there always just one range of slots, according to the name of the function?

torwig commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+// TODO: maybe it needs to use a more precise error type to represent `NotOk`.
+Status Cluster::parseSlotRange(const std::string &slots_str, std::vector<std::pair<int, int>> &slots) {
+  if (slots_str.empty()) {
+    return {Status::NotOK, "Don't use empty slots."};
```
Here you can simply say "No slots to parse" or something like that.

torwig commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+  // Parse all slots(include slot ranges)
+  for (const auto &slot_range : slot_ranges) {
+    if (is_number(slot_range)) {
+      int s_start = stoi(slot_range);
```
Here you can use the `ParseInt` function from `parse_util.h`.

torwig commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    if (is_number(fields[0]) && is_number(fields[1])) {
+      int s_start = stoi(fields[0]), s_end = stoi(fields[1]);
+      assert((s_start <= s_end) && IsValidSlot(s_end));
+      slots.emplace_back(std::make_pair(s_start, s_end));
```
Did you forget to insert `continue` after `emplace_back`?

torwig commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+
+  // Parse all slots(include slot ranges)
+  for (const auto &slot_range : slot_ranges) {
+    if (is_number(slot_range)) {
```
Personally, here I would split the input with `-` and then decide if it's a range or just a single slot number by checking the result.
Because right now you are checking all the characters, then splitting the string by `-` and checking all the characters of each substring.

infdahai replied:
yeah, I change the logic.

mapleFU commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    return {Status::NotOK, fmt::format("Invalid slots: {}. Please use ' ' to space slots.", slots_str)};
+  }
+
+  auto is_number = [&](const std::string &s) {
```
Why not use `src/common/parse_util.h`?

torwig commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    // parse slot range: "int1-int2" (satisfy: int1 <= int2 )
+    assert(slot_range.back() != '-');
+    std::vector<std::string> fields = util::Split(slot_range, "-");
+    assert(fields.size() == 2);
```
Correct me if I'm wrong. `assert` works only in DEBUG build. Thus, in release build, if there is no `-` inside the string, the next line will try to access `fields[`]` and the service will crash. 
In my opinion, such cases should be handled explicitly and return Status::NotOK instead of excessively using `assert`.
However, I would like to hear @PragmaTwice on that as an expert.

infdahai replied:
you're right. I will remove `assert`.

mapleFU commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    }
+
+    // parse slot range: "int1-int2" (satisfy: int1 <= int2 )
+    assert(slot_range.back() != '-');
```
If user pass a ambigious string, would this causing fatal?

mapleFU commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +649,54 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    // parse slot range: "int1-int2" (satisfy: int1 <= int2 )
+    assert(slot_range.back() != '-');
+    std::vector<std::string> fields = util::Split(slot_range, "-");
+    assert(fields.size() == 2);
```
ditto

torwig commented on src/commands/cmd_cluster.cc:
```diff
@@ -219,7 +210,7 @@ class CommandClusterX : public Commander {
       }
     } else if (subcommand_ == "setslot") {
-      Status s = svr->cluster->SetSlot(slot_id_, args_[4], set_version_);
+      Status s = svr->cluster->SetSlot(slots_str_, args_[4], set_version_);
```
Maybe, it would be better to parse slot ranges inside `CommandClusterX::Parse` and then pass already parsed (and valid) numbers to the `SetSlot` function.

infdahai replied:
I'm not sure to do this. `CommandClusterX::Parse` is only used to find some keywords and do the simple check now.
If necessary, I can migrate `parse` functions in `cluster.h` to `CommandClusterX`.

caipengbo replied:
I also think it would be better to do this in command parsing. Only valid slotranges is concerned in `cluster.cc` and these parsing aspects are not concerned.

PragmaTwice commented on src/common/encoding.h:
```diff
@@ -23,6 +23,7 @@
 #include <rocksdb/slice.h>
 #include <unistd.h>
 
+#include <cstdint>
```
```suggestion
```

PragmaTwice commented on CMakeLists.txt:
```diff
@@ -164,6 +166,12 @@ if ((PROJECT_VERSION STREQUAL "unstable") AND (GIT_SHA STREQUAL ""))
 endif ()
 configure_file(src/version.h.in ${PROJECT_BINARY_DIR}/version.h)
 
+if (NOT CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE STREQUAL "None")
+    set (CMAKE_BUILD_TYPE "RelWithDebInfo")
+    message (STATUS "CMAKE_BUILD_TYPE is not set, set to default = ${CMAKE_BUILD_TYPE}")
+endif ()
+message (STATUS "CMAKE_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")
```
I think change the value of `CMAKE_BUILD_TYPE` after the `project()` command (and even after all dependency fetching and configuration here) can be a totally mess (the dependency and many target can be mis-configured).

infdahai replied:
yes, you are right. I just want to make `debug`,so shall I do this check in `x.py build`?

PragmaTwice replied:
IIRC `x.py` already has a default build type.

infdahai replied:
OK, I will remove this part.

PragmaTwice commented on CMakeLists.txt:
```diff
@@ -33,6 +33,8 @@ option(PORTABLE "build a portable binary (disable arch-specific optimizations)"
 # TODO: set ENABLE_NEW_ENCODING to ON when we are ready
 option(ENABLE_NEW_ENCODING "enable new encoding (#1033) for storing 64bit size and expire time in milliseconds" OFF)
 
+set(CMAKE_LINK_DEPENDS_NO_SHARED 1)
```
```suggestion
set(CMAKE_LINK_DEPENDS_NO_SHARED ON)
```

And, could you share some reason for this change?

infdahai replied:
just avoid relinking all depended targets on .so. It's not necessary now, so I delete it.

PragmaTwice commented on x.py:
```diff
@@ -106,7 +106,7 @@ def build(dir: str, jobs: Optional[int], ghproxy: bool, ninja: bool, unittest: b
     makedirs(dir, exist_ok=True)
 
-    cmake_options = ["-DCMAKE_BUILD_TYPE=RelWithDebInfo"]
+    cmake_options = []
```
related to the change to CMakeLists.txt

PragmaTwice commented on src/cluster/redis_slot.h:
```diff
@@ -19,6 +19,7 @@
  */
 
 #pragma once
+#include <cstdint>
```
```suggestion
```

infdahai replied:
Please see the result in https://godbolt.org/z/5q6zj5W9s . So I think the header is due.

PragmaTwice replied:
Got it. Thanks for your demo.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +648,56 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+  }
+
+  auto valid_range = NumericRange<int>{0, kClusterSlots - 1};
+  // Parse all slots(include slot ranges)
```
```suggestion
  // Parse all slots (include slot ranges)
```

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +648,56 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+                  "Invalid slot range: {}. The slot range `int1-int2` needs to satisfy the condition (int1 <= int2).",
+                  slot_range)};
+    }
+    int slot_start = *parse_start;
+    int slot_end = *parse_end;
+    slots.emplace_back(std::make_pair(slot_start, slot_end));
```
```suggestion
    slots.emplace_back(*parse_start, *parse_end);
```

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +648,56 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+    }
+
+    // parse slot range: "int1-int2" (satisfy: int1 <= int2 )
+    if (slot_range.back() == '-') {
+      return {Status::NotOK,
+              fmt::format("Invalid slot range: {}. The character '-' can't appear in the last position.", slot_range)};
+    }
```
It seems to be unnecessary since the code below (split according to `-` and expect `size() == 2`) already did this check.

infdahai replied:
in util/Split(),this process is below.
```cpp
  size_t begin = 0, end = in.find_first_of(delim);
  do {
    std::string elem = in.substr(begin, end - begin);
    if (!elem.empty()) out.push_back(std::move(elem));
    if (end == std::string::npos) break;
    begin = end + 1;
    end = in.find_first_of(delim, begin);
  } while (true);
```
when I use the range such as `3-4-`, I also get a valid result. Because the last `-` just ignores and breaks in this func.

PragmaTwice replied:
Thanks. Got your point.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -128,30 +122,43 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
-      imported_slots_.erase(slot);
+  // remember: The atomicity of the process is based on
+  // the transactionality of ClearKeysOfSlot().
+  for (const auto &slot_range : slots) {
+    int s_start = slot_range.first, s_end = slot_range.second;
```
```suggestion
  for (auto [s_start, s_end] : slots) {
```

PragmaTwice reviewed (commented):
Thanks for your contribution!

You can run `./x.py format` to format the code and let the CI proceed.

infdahai commented:
> Thanks for your contribution!
> 
> You can run `./x.py format` to format the code and let the CI proceed.

If I exec `./x.py format`, I get a irrelevant result.(Not important though, just ignored).
![image](https://user-images.githubusercontent.com/30547248/235607033-0051de12-08f9-499a-830a-a3e3ddc1879e.png)

PragmaTwice commented:
> If I exec `./x.py format`, I get a irrelevant result.(Not important though, just ignored). ![image](https://user-images.githubusercontent.com/30547248/235607033-0051de12-08f9-499a-830a-a3e3ddc1879e.png)

Could you provide the version of clang-format you are using?

Anyway, I think you can firstly manually ignore this unrelated change as a workaround.

PragmaTwice commented:
Sorry for the wrong close-button click. :rofl:

infdahai commented:
> > If I exec `./x.py format`, I get a irrelevant result.(Not important though, just ignored). ![image](https://user-images.githubusercontent.com/30547248/235607033-0051de12-08f9-499a-830a-a3e3ddc1879e.png)
> 
> Could you provide the version of clang-format you are using?
> 
> Anyway, I think you can firstly manually ignore this unrelated change as a workaround.

yes, I will propose a new pr for this `clang-tidy`.

PragmaTwice commented:
CI seems still blocked in format. Not sure your clang-format version, but you can try clang-format 12, which is used in CI.

https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1

And I can confirm that clang-format 14 on my local has also no issue.

```
/home/runner/work/incubator-kvrocks/incubator-kvrocks/src/cluster/cluster.cc:474:39: error: code should be clang-formatted [-Wclang-format-violations]
  for (const auto &id : n->replicas) {      // replicas
                                      ^
```

infdahai commented:
> CI seems still blocked in format. Not sure your clang-format version, but you can try clang-format 12, which is used in CI.
> 
> https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1
> 
> And I can confirm that clang-format 14 on my local has also no issue.
> 
> ```
> /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/cluster/cluster.cc:474:39: error: code should be clang-formatted [-Wclang-format-violations]
>   for (const auto &id : n->replicas) {      // replicas
>                                       ^
> ```

Sorry for this problem. My llvm toolchain version is not unified and the `clang-format` version is 16.

You're too kind. Ignore this problem.

git-hulk commented on src/cluster/cluster.cc:
```diff
@@ -93,32 +93,26 @@ Status Cluster::SetNodeId(const std::string &node_id) {
-Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_version) {
-  // Parameters check
+Status Cluster::SetSlot(const std::string &slots_str, const std::string &node_id, int64_t new_version) {
+  // the parsing process takes some time, so let's check other arguments first.
```
Can remove meaningless comment line
```suggestion
```

git-hulk commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +647,53 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
 }
 
+// TODO: maybe it needs to use a more precise error type to represent `NotOk`.
+Status Cluster::parseSlotRanges(const std::string &slots_str, std::vector<std::pair<int, int>> &slots) {
```
Could you add some test cases for parseSlotRanges?

git-hulk commented on src/cluster/cluster.cc:
```diff
@@ -641,6 +647,53 @@ Status Cluster::LoadClusterNodes(const std::string &file_path) {
+  std::vector<std::string> slot_ranges = util::Split(slots_str, " ");
+
+  if (slot_ranges.empty()) {
+    return {Status::NotOK, fmt::format("Invalid slots: {}. Please use ' ' to space slots.", slots_str)};
```
```suggestion
    return {Status::NotOK, fmt::format("Invalid slots: {}. Please use spaces to separate slots.", slots_str)};
```

PragmaTwice reviewed (commented):
LGTM generally, it would be better if you add several golang tests to tests/gocase/integration/cluster/cluster_test.go .

caipengbo commented on src/cluster/cluster.cc:
```diff
@@ -93,32 +93,26 @@ Status Cluster::SetNodeId(const std::string &node_id) {
 // updates since of network failure, it is state instead of operation.
-Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_version) {
-  // Parameters check
+Status Cluster::SetSlot(const std::string &slots_str, const std::string &node_id, int64_t new_version) {
```
Just pass in the valid slotrange and hand the parsing and validation to the outer layer. You can rename this function to `SetSlots` or `SetSlotRanges`.

infdahai committed:
add stdint.h for uint_t type

1. ref(https://godbolt.org/z/5q6zj5W9s)

Signed-off-by: clundro <859287553@qq.com>

caipengbo commented on src/cluster/cluster.cc:
```diff
@@ -93,32 +83,27 @@ Status Cluster::SetNodeId(const std::string &node_id) {
 // updates since of network failure, it is state instead of operation.
-Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_version) {
-  // Parameters check
+Status Cluster::SetSlotRanges(const std::vector<std::pair<int, int>> &slot_ranges, const std::string &node_id,
```
Personally, I think it's better to use alias declaration(Like `using SlotRange = std::pair<int, int>` ) or struct, because I see `std::pair<int, int>` appear 6 times.

infdahai replied:
OK, I use SlotRange instead of std::pair<int,int>.
And it lefts the `start` and `end` integers in SlotInfo struct. Maybe we think about it later. 

```cpp
// TODO: replace integer `start`, `end` with SlotRange
struct SlotInfo {
  int start;
  int end;
  struct NodeInfo {
    std::string host;
    int port;
    std::string id;
  };
  std::vector<NodeInfo> nodes;
};
```

PragmaTwice commented on src/cluster/cluster_defs.h:
```diff
@@ -0,0 +1,43 @@
+inline constexpr const char *errInvalidClusterNodeInfo = "Invalid cluster nodes info";
+inline constexpr const char *errInvalidImportState = "Invalid import state";
+
+using SlotRange = std::pair<int, int>;
```
```suggestion
using SlotRange = std::pair<int, int>;

```
better to have a newline in the end of the file.

infdahai replied:
I add "\n" to the line in local env but it doesn't seem to see it in github. If I add two `\n` with increasing two line nums and then I get the above situation, but `clang-format` will fix this.

for example,
https://github.com/apache/incubator-kvrocks/blob/30e9fd788a4acb585566a008f4d5fff65ee06ce3/src/cluster/slot_import.h#L53-L59

infdahai commented on src/cluster/cluster_defs.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "cluster/redis_slot.h"
+
+enum {
```
I move the enum struct and strings defined in `cluster` class.
It reduces dependencies and we can add this header in another module, such as `commander.h`.
Perhaps we can move many strings like this if we want to use these in other classs.

infdahai commented on src/cluster/cluster.cc:
```diff
@@ -135,23 +120,29 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+      if (old_node == myself_ && old_node != to_assign_node) {
+        // If slot is migrated from this node
+        if (migrated_slots_.count(slot) > 0) {
+          svr_->slot_migrator->ClearKeysOfSlot(kDefaultNamespace, slot);
```
One `ClearKeyOfSlot` call one writebatch of rocksdb.
So this process exists consistency issues if `slot` in the middle is down.
I don't know what to do with it yet

caipengbo replied:
There is indeed this problem and we can persist it to Propagate CF.

caipengbo replied:
After we merge this PR, you can ask an issue and I will solve this problem, which I have solved in production environment.

git-hulk commented on src/cluster/cluster.cc:
```diff
@@ -93,32 +84,27 @@ Status Cluster::SetNodeId(const std::string &node_id) {
-//
+// Set the slots to the node if new version is current version +1. It is useful
+// when we scale cluster avoid too many big messages. By th way,there are 16384 slots
+// in our design.
```
```suggestion
```

git-hulk commented:
LGTM, I think we can remove TODO comments, then submit an issue if you feel it's necessary.

infdahai commented:
> LGTM, I think we can remove TODO comments, then submit an issue if you feel it's necessary.

OK

infdahai committed:
support cluster_set_slots batch

                1. add  cluster_slot_batch ref to #529

                2. add tests for parsedSlotRange

                3. move parse func to commandX

Signed-off-by: clundro <859287553@qq.com>

caipengbo reviewed (approved):
LGTM

torwig reviewed (approved):
LGTM

git-hulk commented:
Thanks all, merging...
